### PR TITLE
[SPARK-4949]shutdownCallback in SparkDeploySchedulerBackend should be enclosed by synchronized block.

### PR DESCRIPTION
**Pull request by sarutak**
A variable `shutdownCallback` in SparkDeploySchedulerBackend can be accessed from multiple threads so it should be enclosed by synchronized block.

### CODE REVIEW

**Comment by sarutak** [comment]
retest this please.

**Comment by srowen** [comment]
This looks good to me. Most of it is making some fields private that look like they should be, simplifying one synchronization primitive, and does fix a theoretical concurrency problem. Tests pass. I will merge soon if there are no objections.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala [hunk 1]
```diff
@@ -31,16 +34,16 @@ private[spark] class SparkDeploySchedulerBackend(
   with AppClientListener
   with Logging {
 
-  var client: AppClient = null
-  var stopping = false
-  var shutdownCallback : (SparkDeploySchedulerBackend) => Unit = _
-  @volatile var appId: String = _
+  private var client: AppClient = null
+  private var stopping = false
+  private val shutdownCallback: AtomicReference[(SparkDeploySchedulerBackend) => Unit] =
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala]: @sarutak  would this not be simpler as a `@volatile` reference? that's how `appId` is handled below.

**Comment by sarutak** [comment]
@srowen Thank you for picking up this PR. I've changed the `shutdownCallback` to be a simple volatile variable.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala [hunk 2]
```diff
@@ -148,19 +152,16 @@ private[spark] class SparkDeploySchedulerBackend(
       super.applicationId
     }
 
+  def setShutdownCallback(f: SparkDeploySchedulerBackend => Unit) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala]: OK, but why do you need this setter now?

Comment (sarutak) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala]: It's no longer needed so I remove it soon.

**Pull request closed without merging**